src/main/kotlin/com/exactpro/th2/infraoperator/util/RabbitMQUtils.kt

```diff
@@ -0,0 +1,161 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:JvmName("RabbitMQUtils")
+
+package com.exactpro.th2.infraoperator.util
+
+import com.exactpro.th2.infraoperator.configuration.ConfigLoader
+import com.exactpro.th2.infraoperator.model.box.mq.QueueConfiguration
+import com.exactpro.th2.infraoperator.model.box.mq.factory.MessageRouterConfigFactory
+import com.exactpro.th2.infraoperator.model.box.mq.factory.MessageRouterConfigFactoryBox
+import com.exactpro.th2.infraoperator.model.box.mq.factory.MessageRouterConfigFactoryEstore
+import com.exactpro.th2.infraoperator.model.box.mq.factory.MessageRouterConfigFactoryMstore
+import com.exactpro.th2.infraoperator.spec.Th2CustomResource
+import com.exactpro.th2.infraoperator.spec.estore.Th2Estore
+import com.exactpro.th2.infraoperator.spec.mstore.Th2Mstore
+import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.mq.RabbitMQContext
+import com.rabbitmq.client.Channel
+import com.rabbitmq.http.client.domain.ExchangeInfo
+import com.rabbitmq.http.client.domain.QueueInfo
+import io.fabric8.kubernetes.client.KubernetesClient
+import mu.KotlinLogging
+import java.io.IOException
+
+private val K_LOGGER = KotlinLogging.logger { }
+
+fun deleteRabbitMQRubbish() {
+    try {
+        val config = ConfigLoader.loadConfiguration()
+
+        if (!config.rabbitMQManagement.cleanUpOnStart) {
+            K_LOGGER.info { "Cleanup RabbitMQ before start is skipped by config" }
+            return
+        }
+
+        val namespacePrefixes = config.namespacePrefixes
+        val topicExchange = RabbitMQContext.getTopicExchangeName()
+
+        createKubernetesClient().use { kuClient ->
+            val resourceHolder = kuClient.collectRabbitMQRubbish(
+                namespacePrefixes,
+                topicExchange,
+                RabbitMQContext.getTh2Queues(),
+                RabbitMQContext.getTh2Exchanges(),
+            )
+            K_LOGGER.info { "RabbitMQ rubbish: $resourceHolder" }
+            deleteRabbitMQRubbish(resourceHolder, RabbitMQContext::getChannel)
+        }
+    } catch (e: Exception) {
+        K_LOGGER.error(e) { "Delete RabbitMQ rubbish failure" }
+    }
+}
+
+internal fun KubernetesClient.collectRabbitMQRubbish(
+    namespacePrefixes: Set<String>,
+    topicExchange: String,
+    th2Queues: Collection<QueueInfo>,
+    th2Exchanges: Collection<ExchangeInfo>,
+): ResourceHolder = ResourceHolder().apply {
+    val namespaces: Set<String> = namespaces(namespacePrefixes)
+    val factories: Map<Class<out Th2CustomResource>, MessageRouterConfigFactory> = createFactories()
+
+    th2Queues.asSequence()
+        .map(QueueInfo::getName)
+        .forEach(queues::add)
+
+    th2Exchanges.asSequence()
+        .map(ExchangeInfo::getName)
+        .forEach(exchanges::add)
+
+    K_LOGGER.debug { "Actual set in RabbitMQ, queues: $queues, exchanges: $exchanges" }
+
+    if (isHolderEmpty() || namespaces.isEmpty()) {
```

Comment:
note: The namespace check can be moved closer to the start of the method (right after you collect the namespaces)

Reply:
We couldn't do that because code above gets th2 queues and exchanges from RabbitMQ, they are rubbish in namespace case

Reply:
Got it, thank you for explaining this. In my opinion, it would be clearer if we split this function into two - the one that collects the information about queues and exchanges and another one that filters the live queue and exchanges from the collected data. This is just a suggestion - you are free to ignore it